File: Tools/expert_layout_probe.py
#!/usr/bin/env python3
"""Qualified actual-offset expert layout component experiment; no runtime adoption.

Build a bounded four-layer derived artifact, verify it byte for byte against
the original tensor pieces, then atomically publish the complete directory.
Original checkpoint files are always read-only. No global cache purge.
"""
import argparse
import copy
import fcntl
import hashlib
import json
import math
import os
from pathlib import Path
import shutil
import statistics
import struct
import subprocess
import tempfile
import time

from prefill_bench import vm_snapshot, host_conditions, run_child

ROOT = Path(__file__).resolve().parent.parent
LAYERS = [0, 15, 31, 47]
PIECES = [f'{proj}_proj.{part}' for proj in ['gate', 'up', 'down'] for part in ['weight', 'scales', 'biases']]
RECORD_BYTES = 2_764_800


def configure_io(fd, controls, writing=False):
    if controls:
        fcntl.fcntl(fd,controls['no_cache_ext'] if writing else controls['no_cache'],1)
        fcntl.fcntl(fd,controls['read_ahead'],0)


def digest(path, controls=None):
    h = hashlib.sha256()
    with Path(path).open('rb') as f:
        configure_io(f.fileno(), controls)
        for data in iter(lambda: f.read(1024 * 1024), b''): h.update(data)
    return h.hexdigest()


def exact_read(fd, count, offset, read=os.pread):
    if min(count, offset) < 0: raise ValueError('negative read range')
    chunks = []; done = 0
    while done < count:
        try: data = read(fd, count - done, offset + done)
        except InterruptedError: continue
        if not data: raise EOFError('incomplete expert piece')
        if len(data) > count - done: raise ValueError('reader over-reported progress')
        chunks.append(data); done += len(data)
    return b''.join(chunks)


def exact_write(fd, data, write=os.write):
    remaining = memoryview(data)
    while remaining:
        try: count = write(fd, remaining)
        except InterruptedError: continue
        if count <= 0 or count > len(remaining): raise OSError('invalid write progress')
        remaining = remaining[count:]


def source_refs(model):
    headers = {}
    for path in sorted(model.glob('*.safetensors')):
        with path.open('rb') as f:
            length = struct.unpack('<Q', f.read(8))[0]
            if not 0 < length <= 100_000_000: raise ValueError('header length')
            raw = f.read(length)
            if len(raw) != length: raise ValueError('short header')
        for name, value in json.loads(raw).items():
            name = name.removeprefix('language_model.')
            if name in headers: raise ValueError('duplicate tensor')
            if name != '__metadata__': headers[name] = (path, length + 8, value)
    refs = []
    for layer in LAYERS:
        row = []
        for p, piece in enumerate(PIECES):
            path, base, tensor = headers[f'model.layers.{layer}.mlp.switch_mlp.{piece}']
            shape = ([512, 640, 320 if p % 3 == 0 else 40] if p < 6
                     else [512, 2560, 80 if p % 3 == 0 else 10])
            dtype = 'U32' if p % 3 == 0 else 'BF16'
            rb = 819200 if p % 3 == 0 else 51200
            start, end = tensor['data_offsets']
            stat = path.stat()
            if tensor['shape'] != shape or tensor['dtype'] != dtype or not 0 <= start <= end <= stat.st_size-base:
                raise ValueError('tensor shape/dtype/range')
            if end-start != rb*512: raise ValueError('tensor byte count')
            if '\n' in str(path): raise ValueError('newline in model path')
            row.append({'path': str(path.resolve()), 'offset': base+start, 'row_bytes': rb,
                        'source_size': stat.st_size, 'source_mtime_ns': stat.st_mtime_ns,
                        'dtype': dtype, 'shape': shape, 'tensor': f'model.layers.{layer}.mlp.switch_mlp.{piece}'})
        refs.append(row)
    return refs


def verify_sources(refs):
    for row in refs:
        for ref in row:
            st = Path(ref['path']).stat()
            if (st.st_size, st.st_mtime_ns) != (ref['source_size'], ref['source_mtime_ns']):
                raise ValueError('checkpoint changed since artifact construction')


def build_artifact(destination, refs, experts=512, fault=None, controls=None):
    """fault(stage,index) is a bounded interruption/disk-full seam for tests."""
    if destination.exists(): raise FileExistsError(destination)
    verify_sources(refs)
    temp = Path(tempfile.mkdtemp(prefix='.expert-layout-', dir=destination.parent))
    fds = {}; copies = {}; published_refs = copy.deepcopy(refs)
    try:
        for row in refs:
            for ref in row:
                if ref['path'] not in fds:
                    fds[ref['path']] = os.open(ref['path'], os.O_RDONLY)
                    configure_io(fds[ref['path']],controls)
                    if controls:
                        name=f'source-{len(copies):03}.bin'
                        fd=os.open(temp/name,os.O_CREAT|os.O_EXCL|os.O_RDWR,0o600)
                        copies[ref['path']]={'fd':fd,'name':name}
                        configure_io(fd,controls,writing=True)
                        # Preserve the original tensor offsets and file extent;
                        # unselected regions remain holes and are never read.
                        os.ftruncate(fd,ref['source_size'])
        packed = temp/'experts.bin'
        fd = os.open(packed, os.O_CREAT | os.O_EXCL | os.O_WRONLY, 0o600)
        h = hashlib.sha256(); size = 0; count = 0; record_digests = bytearray()
        try:
            configure_io(fd,controls,writing=True)
            for row in refs:
                for e in range(experts):
                    record_hash = hashlib.sha256()
                    for ref in row:
                        data = exact_read(fds[ref['path']], ref['row_bytes'], ref['offset']+e*ref['row_bytes'])
                        if controls:
                            offset=ref['offset']+e*ref['row_bytes']; progress=0
                            def write_copy(unused, part):
                                nonlocal progress
                                wrote=os.pwrite(copies[ref['path']]['fd'],part,offset+progress)
                                progress+=wrote
                                return wrote
                            exact_write(0,data,write_copy)
                        exact_write(fd, data); h.update(data); record_hash.update(data); size += len(data)
                    record_digests.extend(record_hash.digest())
                    if fault: fault('write', count)
                    count += 1
            os.fsync(fd)
        finally: os.close(fd)
        for item in copies.values(): os.fsync(item['fd'])
        # Re-read the finished file and every original piece independently;
        # a digest of bytes sent to write is not proof of bytes stored.
        with packed.open('rb') as check:
            configure_io(check.fileno(),controls)
            for row in refs:
                for e in range(experts):
                    for ref in row:
                        original = exact_read(fds[ref['path']], ref['row_bytes'], ref['offset']+e*ref['row_bytes'])
                        if check.read(len(original)) != original: raise ValueError('derived bytes differ')
                        if controls and exact_read(copies[ref['path']]['fd'],ref['row_bytes'],ref['offset']+e*ref['row_bytes']) != original:
                            raise ValueError('column-copy bytes differ')
            if check.read(1): raise ValueError('derived trailing data')
        if packed.stat().st_size != size or digest(packed,controls) != h.hexdigest(): raise ValueError('stored digest/length')
        verify_sources(refs)
        if controls:
            for row in published_refs:
                for ref in row:
                    item=copies[ref['path']]; stat=os.fstat(item['fd'])
                    ref.update(path=str(destination/item['name']),source_size=stat.st_size,source_mtime_ns=stat.st_mtime_ns)
        with (temp/'record-digests.bin').open('xb') as f:
            f.write(record_digests); f.flush(); os.fsync(f.fileno())
        manifest = {'version': 1, 'layout': 'layer-expert-piece-v1', 'refs': published_refs,
                    'checkpoint_refs':refs,'io_controls':controls,
                    'experts_per_layer': experts, 'bytes': size, 'sha256': h.hexdigest(),
                    'record_digests_sha256':hashlib.sha256(record_digests).hexdigest(),
                    'record_digests_bytes':len(record_digests),
                    'physical_allocated_bytes':sum(p.stat().st_blocks*512 for p in temp.iterdir() if p.is_file())}
        with (temp/'manifest.json').open('x') as f:
            json.dump(manifest, f, indent=2); f.write('\n'); f.flush(); os.fsync(f.fileno())
        if fault: fault('publish', count)
        os.rename(temp, destination)
        directory = os.open(destination.parent, os.O_RDONLY)
        try: os.fsync(directory)
        finally: os.close(directory)
        return manifest
    finally:
        for fd in fds.values(): os.close(fd)
        for item in copies.values(): os.close(item['fd'])
        if temp.exists(): shutil.rmtree(temp)


def validate_artifact(path):
    m = json.loads((path/'manifest.json').read_text())
    if m.get('version') != 1 or m.get('layout') != 'layer-expert-piece-v1': raise ValueError('layout identity')
    size = sum(sum(r['row_bytes'] for r in row) for row in m['refs']) * m['experts_per_layer']
    if m['bytes'] != size or (path/'experts.bin').stat().st_size != size: raise ValueError('layout size')
    if digest(path/'experts.bin',m.get('io_controls')) != m['sha256']: raise ValueError('layout digest')
    if 'record_digests_sha256' in m:
        hashes=path/'record-digests.bin'
        if hashes.stat().st_size != len(m['refs'])*m['experts_per_layer']*32 or digest(hashes)!=m['record_digests_sha256']:
            raise ValueError('record digest identity')
    verify_sources(m['refs'])
    if 'checkpoint_refs' in m: verify_sources(m['checkpoint_refs'])
    return m


def summarize(cells, protocol):
    results = []
    for workload in range(6):
        pairs = []
        for round_id in range(1, protocol['rounds']+1):
            pair = [x for x in cells if x['workload'] == workload and x['round'] == round_id]
            if len(pair) != 2 or any(x['exclusions'] for x in pair): continue
            ref = next(x['result'] for x in pair if x['mode'] == 0)
            cand = next(x['result'] for x in pair if x['mode'] != 0)
            same = all(ref[k] == cand[k] for k in ['records', 'batches', 'bytes', 'tensor_sha256'])
            pairs.append({'round': round_id, 'exact': same, 'reduction': 1-cand['seconds']/ref['seconds'],
                          'reference_seconds': ref['seconds'], 'packed_seconds': cand['seconds']})
        median = statistics.median(x['reduction'] for x in pairs) if pairs else None
        passed = len(pairs) >= protocol['minimum_pairs'] and all(x['exact'] for x in pairs)
        passed = passed and median >= protocol['minimum_median_reduction']
        passed = passed and sum(x['reduction'] > 0 for x in pairs)/len(pairs) >= protocol['minimum_positive_fraction']
        results.append({'workload': workload, 'name': protocol['workloads'][workload], 'pairs': pairs,
                        'median_reduction': median, 'passed': bool(passed)})
    return results


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--model', type=Path, required=True)
    parser.add_argument('--out', type=Path, required=True)
    parser.add_argument('--verify-records',action='store_true')
    args = parser.parse_args(); out = args.out.resolve(); out.mkdir(parents=True, exist_ok=False)
    protocol = {'kind': 'actual-offset component qualification; no model speed claim',
        'layers': LAYERS, 'rounds': 7, 'minimum_pairs': 5, 'minimum_median_reduction': .05,
        'minimum_positive_fraction': .8, 'required_reclaimable_gb': 4,
        'maximum_process_bytes': 512_000_000, 'maximum_buffer_bytes': 512_000_000,
        'workloads': ['random pool 1 miss', 'random pool 4 misses', 'random pool 10 misses',
                      'random pool 32 misses', 'dense sweep 32', 'half-density sweep 32'],
        'timing': 'allocation, job scheduling, uncached reads, packed deinterleave, and free; excludes equal full-payload SHA-256 verification',
        'reference': 'actual tensor offsets; pool 32 static lanes; sweep 12 longest-first dynamic lanes',
        'candidate': 'verified packed records; same output columns; per-lane aligned scratch; pool 32 / sweep 12 lanes',
        'minimum_disk_read_fraction': .98,
        'exclusions': 'any global swap change over a cell excludes its whole pair; disk-read bytes below 98% of requested payload; missing/invalid metrics or failed bytes; no replacements',
        'scope': 'four complete selected layers; packed payload 5,662,310,400 bytes plus uncached sparse column copies at original checkpoint offsets (same selected payload); original weights stay read-only; no runtime loading or direct GPU writes'}
    candidate_mode=2 if args.verify_records else 1
    protocol['candidate_mode']=candidate_mode
    protocol['record_verification']='SHA-256 of every packed record is compared with the construction digest inside the timed worker' if args.verify_records else 'artifact verified before/after study; output SHA outside timing'
    for name in ['expert_layout_probe.py','expert_layout_probe.c','prefill_bench.py']:
        shutil.copy2(ROOT/'Tools'/name,out/name)
    protocol['source_hashes'] = {p.name: digest(p) for p in out.iterdir() if p.is_file()}
    (out/'protocol.json').write_text(json.dumps(protocol,indent=2)+'\n')
    manifest = {'protocol': protocol, 'completed': False, 'cells': []}
    with open(f'/tmp/slotstream-model-{os.getuid()}.lock','a') as lock:
        # Acquire and retain the same lock before disk work. This experiment
        # does not instantiate a model and never races a model benchmark.
        fcntl.flock(lock,fcntl.LOCK_EX|fcntl.LOCK_NB)
        before = vm_snapshot(); manifest['before'] = before
        if before['reclaimable_bytes'] < 4e9: raise RuntimeError('insufficient headroom')
        if shutil.disk_usage(out).free < 16e9: raise RuntimeError('need 16 GB disk headroom')
        try:
            command = ['cc','-O2','-std=c11','-Wno-deprecated-declarations',str(out/'expert_layout_probe.c'),'-o',str(out/'probe')]
            built = subprocess.run(command,capture_output=True,text=True)
            (out/'build.txt').write_text(built.stdout+built.stderr)
            built.check_returncode(); manifest['native_sha256'] = digest(out/'probe')
            manifest['build_command'] = command
            controls=json.loads(subprocess.check_output([str(out/'probe'),'--io-constants'],text=True))
            print('build and byte-verify uncached packed and column artifacts',flush=True)
            checkpoint_refs = source_refs(args.model.resolve())
            manifest['artifact'] = build_artifact(out/'artifact',checkpoint_refs,controls=controls)
            validate_artifact(out/'artifact'); refs=manifest['artifact']['refs']
            config = ['SSLP2',str(out/'artifact/experts.bin'),str(out/'artifact/record-digests.bin')]
            for row in refs:
                for ref in row: config += [ref['path'],f"{ref['offset']} {ref['row_bytes']}"]
            (out/'native-manifest.txt').write_text('\n'.join(config)+'\n')
            for workload in range(6):
                for round_id in range(1,8):
                    for mode in ([0,candidate_mode] if round_id % 2 else [candidate_mode,0]):
                        cell = out/f'w{workload}-r{round_id}-m{mode}'; cell.mkdir()
                        value = {'workload': workload,'round': round_id,'mode': mode,'exclusions':[]}
                        value['before'] = vm_snapshot(); value['conditions_before'] = host_conditions()
                        try:
                            verify_sources(refs)
                            verify_sources(checkpoint_refs)
                            if value['before']['reclaimable_bytes'] < 4e9: raise RuntimeError('insufficient headroom')
                            command=[str(out/'probe'),str(out/'native-manifest.txt'),str(mode),str(workload)]
                            value['command'] = command
                            code = run_child(command,os.environ.copy(),cell,120)
                            value['exit_code']=code
                            if code: raise RuntimeError(f'native exit {code}')
                            result=json.loads((cell/'stdout.txt').read_text()); value['result']=result
                            if type(result['seconds']) not in [int,float] or not math.isfinite(result['seconds']) or result['seconds'] <= 0:
                                raise ValueError('invalid duration')
                            expected_records=1024 if workload==5 else 2048
                            if result['records']!=expected_records or result['bytes']!=expected_records*RECORD_BYTES:
                                raise ValueError('incomplete native work')
                            if max(result['sampled_footprint_bytes'],result['lifetime_rss_peak_bytes']) > protocol['maximum_process_bytes']:
                                value['exclusions'].append('physical footprint over bound')
                            if result['maximum_owned_buffer_bytes'] > protocol['maximum_buffer_bytes']:
                                value['exclusions'].append('owned buffers over bound')
                            if result['diskio_bytesread'] < result['bytes']*protocol['minimum_disk_read_fraction']:
                                value['exclusions'].append('disk reads do not cover requested payload')
                        except Exception as e: value['exclusions'].append(f'{type(e).__name__}: {e}')
                        finally:
                            value['after']=vm_snapshot()
                            if any(value['before'][k] != value['after'][k] for k in ['swapins','swapouts']):
                                value['exclusions'].append('global swap activity')
                            (cell/'result.json').write_text(json.dumps(value,indent=2)+'\n')
                            manifest['cells'].append(value)
                            (out/'manifest.json').write_text(json.dumps(manifest,indent=2)+'\n')
                            print(f"workload {workload} round {round_id} mode {mode}: "
                                  f"{value.get('result',{}).get('seconds')} s; {value['exclusions']}",flush=True)
            manifest['artifact_verified_after']=validate_artifact(out/'artifact') == manifest['artifact']
            manifest['completed']=True
        finally:
            manifest['summary']=summarize(manifest['cells'],protocol)
            manifest['sources_unchanged']=all(digest(ROOT/'Tools'/name)==sha for name,sha in protocol['source_hashes'].items())
            manifest['after']=vm_snapshot()
            (out/'manifest.json').write_text(json.dumps(manifest,indent=2)+'\n')
    print(json.dumps(manifest['summary'],indent=2),flush=True)


if __name__ == '__main__': main()

File: Tools/expert_layout_probe.c
// Bounded, actual-offset layout experiment. No model or GPU allocations.
// Reference mirrors ExpertStore's per-piece pool lanes and longest-first
// sweep jobs. Packed includes conversion back to nine contiguous columns.
#include <CommonCrypto/CommonDigest.h>
#include <dispatch/dispatch.h>
#include <errno.h>
#include <fcntl.h>
#include <libproc.h>
#include <mach/mach.h>
#include <pthread.h>
#include <stdatomic.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/resource.h>
#include <sys/stat.h>
#include <time.h>
#include <unistd.h>

#define PIECES 9
#define LAYERS 4
#define EXPERTS 512
#define MAX_ROWS 32
#define RECORD_BYTES 2764800
typedef struct { int fd; off_t offset; size_t row; } ref_t;
typedef struct { int piece, row, length; size_t bytes; } job_t;
typedef struct {
    ref_t *refs; int packed, mode, sweep, n, lanes, njobs, next;
    off_t packed_base; int *ids; void **out; job_t jobs[PIECES*MAX_ROWS];
    const unsigned char *digests;
    pthread_mutex_t lock; _Atomic int error; _Atomic long calls;
} batch_t;
static double now(void) {
    struct timespec t; clock_gettime(CLOCK_MONOTONIC, &t);
    return t.tv_sec + t.tv_nsec / 1e9;
}
static void die(const char *s) { fprintf(stderr, "%s: %s\n", s, strerror(errno)); exit(2); }
static _Atomic uint64_t sampled_peak=0;
static _Atomic int sampling_stop=0, sampling_error=0;
static void sample_memory(void) {
    struct task_vm_info info; mach_msg_type_number_t cnt=TASK_VM_INFO_COUNT;
    if(task_info(mach_task_self(),TASK_VM_INFO,(task_info_t)&info,&cnt)!=KERN_SUCCESS) {
        atomic_store(&sampling_error,1); return;
    }
    uint64_t prior=atomic_load(&sampled_peak);
    while(prior<info.phys_footprint && !atomic_compare_exchange_weak(&sampled_peak,&prior,info.phys_footprint)) {}
}
static void *sample_loop(void *unused) {
    (void)unused;
    do { sample_memory(); usleep(20000); } while(!atomic_load(&sampling_stop));
    sample_memory(); return NULL;
}
static void *aligned(size_t bytes) {
    void *p = NULL; int e = posix_memalign(&p,16384,bytes);
    if(e) { errno=e; die("allocation"); } return p;
}
static int exact(batch_t *b, int fd, void *dst, size_t bytes, off_t offset) {
    size_t done=0;
    while(done<bytes) {
        if(atomic_load(&b->error)) return 0;
        ssize_t n=pread(fd,(char*)dst+done,bytes-done,offset+done);
        atomic_fetch_add(&b->calls,1);
        if(n<0 && errno==EINTR) continue;
        if(n<=0) { atomic_store(&b->error,n==0?ENODATA:errno); return 0; }
        done+=(size_t)n;
    }
    return 1;
}
static void work(void *context,size_t lane) {
    batch_t *b=context;
    size_t largest=0;
    for(int j=0;j<b->njobs;j++) if(b->jobs[j].bytes>largest) largest=b->jobs[j].bytes;
    void *scratch=b->mode?aligned(largest):NULL;
    int cursor=(int)lane;
    while(!atomic_load(&b->error)) {
        int j=cursor;
        if(b->sweep) { pthread_mutex_lock(&b->lock); j=b->next++; pthread_mutex_unlock(&b->lock); }
        else cursor+=b->lanes;
        if(j>=b->njobs) break;
        job_t job=b->jobs[j];
        if(!b->mode) {
            ref_t ref=b->refs[job.piece];
            if(!exact(b,ref.fd,(char*)b->out[job.piece]+job.row*ref.row,
                job.bytes,ref.offset+(off_t)b->ids[job.row]*ref.row)) break;
        } else {
            if(!exact(b,b->packed,scratch,job.bytes,
                b->packed_base+(off_t)b->ids[job.row]*RECORD_BYTES)) break;
            for(int row=0;row<job.length;row++) {
                if(b->mode==2) {
                    unsigned char hash[CC_SHA256_DIGEST_LENGTH];
                    CC_SHA256((char*)scratch+row*RECORD_BYTES,RECORD_BYTES,hash);
                    if(memcmp(hash,b->digests+b->ids[job.row+row]*CC_SHA256_DIGEST_LENGTH,CC_SHA256_DIGEST_LENGTH)) {
                        atomic_store(&b->error,EBADMSG); break;
                    }
                }
                size_t offset=0;
                for(int p=0;p<PIECES;p++) {
                    size_t bytes=b->refs[p].row;
                    memcpy((char*)b->out[p]+(job.row+row)*bytes,
                        (char*)scratch+row*RECORD_BYTES+offset,bytes);
                    offset+=bytes;
                }
            }
        }
    }
    free(scratch);
}
static int largest_first(const void *a,const void *b) {
    size_t x=((const job_t*)a)->bytes,y=((const job_t*)b)->bytes;
    return (x<y)-(x>y);
}
static int ascending(const void *a,const void *b) { return *(const int*)a-*(const int*)b; }
static int by_time(const void *a,const void *b) {
    double x=*(const double*)a,y=*(const double*)b; return (x>y)-(x<y);
}
static void line(FILE *f,char *out,size_t cap) {
    if(!fgets(out,(int)cap,f)) { errno=EINVAL; die("manifest line"); }
    size_t n=strlen(out);
    if(!n||out[n-1]!='\n') { errno=EINVAL; die("manifest line length"); }
    out[n-1]=0;
}
static int open_data(const char *path) {
    int fd=open(path,O_RDONLY|O_NOFOLLOW);
    if(fd<0) die("open");
    if(fcntl(fd,F_NOCACHE,1)<0||fcntl(fd,F_RDAHEAD,0)<0) die("uncached I/O");
    return fd;
}
static void warm(void *unused,size_t lane) { (void)unused; (void)lane; }
int main(int argc,char **argv) {
    if(argc==2 && !strcmp(argv[1],"--io-constants")) {
        printf("{\"no_cache\":%d,\"no_cache_ext\":%d,\"read_ahead\":%d}\n",F_NOCACHE,F_NOCACHE_EXT,F_RDAHEAD);
        return 0;
    }
    if(argc!=4||strlen(argv[2])!=1||strlen(argv[3])!=1||argv[2][0]<'0'||argv[2][0]>'2'
        ||argv[3][0]<'0'||argv[3][0]>'5') {
        fprintf(stderr,"usage: probe manifest mode(0=pieces,1=packed,2=verified-packed) workload(0..5)\n"); return 2;
    }
    int mode=argv[2][0]-'0', workload=argv[3][0]-'0';
    FILE *file=fopen(argv[1],"r"); if(!file) die("manifest");
    char path[4096]; line(file,path,sizeof(path));
    int version=!strcmp(path,"SSLP2")?2:!strcmp(path,"SSLP1")?1:0;
    if(!version || (mode==2 && version!=2)) { errno=EINVAL; die("manifest version"); }
    line(file,path,sizeof(path)); int packed=open_data(path);
    struct stat st; if(fstat(packed,&st)||st.st_size!=(off_t)LAYERS*EXPERTS*RECORD_BYTES) {
        errno=EINVAL; die("packed length");
    }
    unsigned char digests[LAYERS*EXPERTS*CC_SHA256_DIGEST_LENGTH];
    if(version==2) {
        line(file,path,sizeof(path)); FILE *hashes=fopen(path,"rb");
        if(!hashes) die("record digests");
        if(fread(digests,1,sizeof(digests),hashes)!=sizeof(digests)||fgetc(hashes)!=EOF) {
            errno=EINVAL; die("record digests length");
        }
        fclose(hashes);
    }
    ref_t refs[LAYERS][PIECES];
    for(int l=0;l<LAYERS;l++) {
        size_t total=0;
        for(int p=0;p<PIECES;p++) {
            line(file,path,sizeof(path)); refs[l][p].fd=open_data(path);
            line(file,path,sizeof(path)); long long offset; unsigned long row; char extra;
            if(sscanf(path,"%lld %lu %c",&offset,&row,&extra)!=2||offset<0
                ||row!=(p%3==0?819200:51200)) { errno=EINVAL; die("tensor geometry"); }
            if(fstat(refs[l][p].fd,&st)||offset>st.st_size
                ||(off_t)row*EXPERTS>st.st_size-offset) { errno=EINVAL; die("tensor range"); }
            refs[l][p].offset=offset; refs[l][p].row=row; total+=row;
        }
        if(total!=RECORD_BYTES) { errno=EINVAL; die("record geometry"); }
    }
    if(fgetc(file)!=EOF) { errno=EINVAL; die("manifest trailing data"); } fclose(file);
    int sizes[]={1,4,10,32,32,32}; int batchsize=sizes[workload],sweep=workload>=4;
    dispatch_queue_t queue=dispatch_get_global_queue(QOS_CLASS_USER_INITIATED,0);
    dispatch_apply_f(32,queue,NULL,warm);
    struct rusage_info_v4 io_before={0},io_after={0};
    if(proc_pid_rusage(getpid(),RUSAGE_INFO_V4,(rusage_info_t*)&io_before)) die("disk I/O observation");
    CC_SHA256_CTX checksum; CC_SHA256_Init(&checksum);
    double seconds=0,times[LAYERS*EXPERTS]; int batches=0,records=0; long calls=0;
    size_t max_owned=0; pthread_t sampler;
    int launch=pthread_create(&sampler,NULL,sample_loop,NULL);
    if(launch) { errno=launch; die("memory sampler"); }
    for(int l=0;l<LAYERS;l++) {
        int ids[EXPERTS]; for(int i=0;i<EXPERTS;i++) ids[i]=i;
        uint32_t rng=0x91e10da5U+(uint32_t)l;
        if(workload!=4) {
            for(int i=EXPERTS-1;i>0;i--) {
                rng^=rng<<13; rng^=rng>>17; rng^=rng<<5;
                int j=(int)(rng%(uint32_t)(i+1)),temp=ids[i]; ids[i]=ids[j];ids[j]=temp;
            }
        }
        int count=workload==5?256:EXPERTS;
        if(sweep) qsort(ids,count,sizeof(int),ascending);
        for(int start=0;start<count;start+=batchsize) {
            int n=count-start<batchsize?count-start:batchsize;
            double t=now(); void *out[PIECES];
            for(int p=0;p<PIECES;p++) out[p]=aligned((size_t)n*refs[l][p].row);
            batch_t b={.refs=refs[l],.packed=packed,.mode=mode,.sweep=sweep,
                .n=n,.packed_base=(off_t)l*EXPERTS*RECORD_BYTES,.ids=ids+start,.out=out,
                .digests=digests+l*EXPERTS*CC_SHA256_DIGEST_LENGTH,
                .lock=PTHREAD_MUTEX_INITIALIZER,.error=0,.calls=0};
            for(int row=0;row<n;) {
                int len=1;
                if(sweep) while(row+len<n && b.ids[row+len]==b.ids[row]+len) len++;
                if(mode) b.jobs[b.njobs++]=(job_t){0,row,len,(size_t)len*RECORD_BYTES};
                else for(int p=0;p<PIECES;p++)
                    b.jobs[b.njobs++]=(job_t){p,row,len,(size_t)len*refs[l][p].row};
                row+=len;
            }
            if(sweep) qsort(b.jobs,b.njobs,sizeof(job_t),largest_first);
            b.lanes=sweep?12:32; if(b.lanes>b.njobs) b.lanes=b.njobs;
            dispatch_apply_f((size_t)b.lanes,queue,&b,work);
            pthread_mutex_destroy(&b.lock);
            double elapsed=now()-t;
            int error=atomic_load(&b.error); if(error) { errno=error; die("batch read"); }
            calls+=atomic_load(&b.calls);
            // Hash every returned tensor byte outside the measured region.
            // Both paths produce exactly the same column ordering.
            for(int p=0;p<PIECES;p++) CC_SHA256_Update(&checksum,out[p],(CC_LONG)((size_t)n*refs[l][p].row));
            sample_memory();
            size_t scratch=0; if(mode) for(int j=0;j<b.njobs;j++) if(b.jobs[j].bytes>scratch) scratch=b.jobs[j].bytes;
            size_t owned=(size_t)n*RECORD_BYTES+scratch*(size_t)b.lanes;
            if(owned>max_owned) max_owned=owned;
            t=now(); for(int p=0;p<PIECES;p++) free(out[p]); elapsed+=now()-t;
            times[batches++]=elapsed; seconds+=elapsed; records+=n;
        }
    }
    atomic_store(&sampling_stop,1); pthread_join(sampler,NULL);
    if(proc_pid_rusage(getpid(),RUSAGE_INFO_V4,(rusage_info_t*)&io_after)) die("disk I/O observation");
    if(atomic_load(&sampling_error)) { errno=EIO; die("footprint sampler"); }
    struct rusage usage; if(getrusage(RUSAGE_SELF,&usage)) die("RSS observation");
    unsigned char hash[CC_SHA256_DIGEST_LENGTH]; CC_SHA256_Final(hash,&checksum);
    qsort(times,batches,sizeof(double),by_time);
    printf("{\"mode\":%d,\"workload\":%d,\"records\":%d,\"batches\":%d,\"bytes\":%lld,"
        "\"read_calls\":%ld,\"seconds\":%.9f,\"batch_median_seconds\":%.9f,"
        "\"batch_p95_seconds\":%.9f,\"sampled_footprint_bytes\":%llu,\"lifetime_rss_peak_bytes\":%ld,"
        "\"sampling_interval_ms\":20,\"maximum_owned_buffer_bytes\":%zu,"
        "\"diskio_bytesread\":%llu,\"reformat_bytes\":%lld,\"tensor_sha256\":\"",
        mode,workload,records,batches,(long long)records*RECORD_BYTES,calls,seconds,
        times[batches/2],times[(batches-1)*95/100],(unsigned long long)atomic_load(&sampled_peak),usage.ru_maxrss,max_owned,
        (unsigned long long)(io_after.ri_diskio_bytesread-io_before.ri_diskio_bytesread),
        mode?(long long)records*RECORD_BYTES:0);
    for(int i=0;i<CC_SHA256_DIGEST_LENGTH;i++) printf("%02x",hash[i]); puts("\"}");
    for(int l=0;l<LAYERS;l++) for(int p=0;p<PIECES;p++) close(refs[l][p].fd);
    close(packed); return 0;
}
